### DISCUSSION TITLE
Custom validators, referencing custom fields and folders

### DISCUSSION THREAD

**Discussion by AndMrzv** [netbox-community/netbox#8950] [Q&A] [answered]
Been experimenting with custom validators and I got two questions:
1) How do I reference a custom field in a custom validator? For example I have a custom field "customfield" (has to be set if the instance status is A or B) attributed to racks but when I use "instance.customfield" in my custom validator I get
`'Rack' object has no attribute 'customfield'`
in web GUI. Tried "cf.customfield", "cf__customfield" as worked in other uses, no luck. Is this feature still limited?
2) Does custom validator has to be inside netbox main directory to be assigned at config revision section or is there a way to use the default path (/etc/netbox/validators, or any other folder)? Attempts at referencing any outlying folders lead me to errors of missing modules.

**Comment by jeremystretch** [comment, marked as answer]
> 1. How do I reference a custom field in a custom validator?

I don't think it's possible using the bare rules approach, but you should be able to do this by creating a [`CustomValidator` class](https://netbox.readthedocs.io/en/stable/customization/custom-validation/#custom-validation-logic). Custom fields can be accessed as `instance.cf.foo`. (Still, it's probably worth an FR to enable custom field access using bare rules.)

> 2. Does custom validator has to be inside netbox main directory to be assigned at config revision section or is there a way to use the default path (/etc/netbox/validators, or any other folder)?

Custom validator classes can be defined anywhere within the Python path. They just need to be importable within `configuration.py` for assignment under the `CUSTOM_VALIDATORS` config parameter.

Hope that helps!

**Reply by AndMrzv** [reply]
Thank you for the reply.
> I don't think it's possible using the bare rules approach, but you should be able to do this by creating a [CustomValidator class](https://netbox.readthedocs.io/en/stable/customization/custom-validation/#custom-validation-logic). Custom fields can be accessed as instance.cf.foo. (Still, it's probably worth an FR to enable custom field access using bare rules.)

This is exactly what I am doing. I have a custom validator assigned via a config revision
`{'dcim.rack': ['validators.test-validator.Myvalidator']}`
to racks for which there is a "customfield" custom field (it shows in API  "custom_fields": {"customfield": null})

```
from extras.validators import CustomValidator

class Myvalidator(CustomValidator):
    def validate(self, instance):
        if not instance.cf.customfield:
            self.fail("test5")
```
And this gets me
![Снимок экрана от 2022-03-28 13-19-58](https://user-images.githubusercontent.com/99031446/160380108-50e0ab86-3f3a-4327-b703-edf63dc3f7fd.png)
Basic fields work fine.

Meanwhile I got a new question. Can I run checks if the value is unique (serial number for example, for same type of objects) or if a some value of a different object matches some conditions (maybe a certain field must be set for a device depending on the role of it's rack)?

**Reply by jeremystretch** [reply]
Ah, sorry. `cf` is actually a dictionary of custom fields and their values, so you want to check `instance.cf["customfield"]`.

**Reply by AndMrzv** [reply]
Yes, now this worked. I used `instance.cf["customfield"]` in the if part and `cf_customfield` as self.fail field value.
Thanks!

**Reply by AndMrzv** [reply]
> Meanwhile I got a new question. Can I run checks if the value is unique (serial number for example, for same type of objects) or if a some value of a different object matches some conditions (maybe a certain field must be set for a device depending on the role of it's rack)?

Figured out how to do this too, all good here.

**Reply by tomehb** [reply]
Took me a few attempts to get this working, so thought I would share:
```
from extras.validators import CustomValidator

class cvMyField(CustomValidator):

    def validate(self, instance):
        from dcim.models import Device
        if instance.cf["myfeild"] and Device.objects.filter(custom_field_data__myfeild=instance.cf["myfeild"]).exists():
            self.fail("Duplicate DTM Address", field='cf_myfeild')

CUSTOM_VALIDATORS = {
    'dcim.device': (
        cvMyField(),
    )
}
```

**Reply by PeterKnotek** [reply]
How define in Tenant / custom filed with name MAC : uniqe MAC adress with format ff:ff:ff:ff:ff:ff oR FF:FF:FF:FF:FF:FF
I have 

from extras.validators import CustomValidator
from tenancy.models import Tenant
import re

class MyValidatorHelper:

    @staticmethod
    def duplicateValidator(needle, haystack):
        fail_message = []
        for hay in haystack:
            if needle.id != hay.id:
                if needle.name.lower() == hay.name.lower():
                    fail_message.append(" Nazov {} je uz pouzity !!".format(hay.name))
        return fail_message

class TenantValidator(CustomValidator):

    def validate(self, instance):
        fail_message = []
        # Format
        if not re.match(r'^[a-zA-Z0-9_.-]*$', instance.name):
            fail_message.append(" ## NEPLATNE ZNAKY ## !! Povolene su == a-z | A-Z | -_ a bodka | medzera je neplatna ==)")
        # DuplicateName
        fail_message += MyValidatorHelper.duplicateValidator(instance, Tenant.objects.all())
        if fail_message:
            self.fail("Nazov pola : {}".format(" and ".join(fail_message)), field='name')

cant duplicate names in Tenenat.name field -  work fine
But how define cf.MAC ... 

Thanks

**Reply by PeterKnotek** [reply]
Solved

Custom filed for tenant = TenantMAC
Code is 

class UniqueTenantMACValidator(CustomValidator):

    def validate(self, instance):
        # Get the value from the TenantMAC custom field for the current instance
        tenant_mac = instance.custom_field_data.get('TenantMAC')

        # Get all tenants that have the same value in the TenantMAC field
        tenants_with_same_mac = Tenant.objects.filter(custom_field_data__TenantMAC=tenant_mac)

        # If there is another tenant with the same MAC address, display an error
        if tenants_with_same_mac.exists() and tenants_with_same_mac.first().id != instance.id:
            self.fail("MAC Addressa uz je pouzita !!", field='cf_TenantMAC')
